File: server_asp/src/api/product/ProductService.cs
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;
using Microsoft.IdentityModel.Tokens;
using shieldtify.common;
using shieldtify.models;
using BC = BCrypt.Net.BCrypt;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
namespace shieldtify.api.product
{
    public static class ProductService
    {
        public static APIRes getAllProduct(int page = 1, string sort = "popular", string priceRange = "", string brandsFilter = "", string keyword = "")
        {
            try
            {
                using var dbContext = new ShieldtifyContext();
                var query = dbContext.Items.Where(i => i.Name.ToLower().Contains(keyword.ToLower()));
                if (priceRange != "")
                {
                    var priceRangeArr = priceRange.Split('-');
                    query = query.Where(i => i.Price >= int.Parse(priceRangeArr[0]) && i.Price <= int.Parse(priceRangeArr[1]));
                }
                if (brandsFilter != "")
                {
                    var brandsArr = brandsFilter.Split(',');
                    query = query.Where(i => brandsArr.Contains(i.Brand.Name));
                }
                if (sort == "popular")
                {
                    query = query.OrderByDescending(i => i.OrderItems.Count());
                }
                else
                {
                    var sortField = sort.Split('-')[0];
                    var sortOrder = sort.Split('-')[1];
                    if (sortField == "price")
                    {
                        query = sortOrder == "asc" ? query.OrderBy(i => i.Price) : query.OrderByDescending(i => i.Price);
                    }
                    else if (sortField == "name")
                    {
                        query = sortOrder == "asc" ? query.OrderBy(i => i.Name) : query.OrderByDescending(i => i.Name);
                    }
                }
                var products = query.Select(i => new
                {
                    uid = i.Uid,
                    name = i.Name,
                    price = i.Price,
                    primary_img = i.ItemImgs.Where(im => im.IsPrimary).Select(im => im.Link).FirstOrDefault(),
                    stock_qty = i.StockQty,
                    brand = i.Brand.Name
                }).Skip((page - 1) * 16).Take(16).ToList();
                // round up max price
                var maxPrice = (int)Math.Ceiling(dbContext.Items.Where(i => i.Name.ToLower().Contains(keyword.ToLower())).Max(i => i.Price));
                var brands = dbContext.Items.Where(i => i.Name.ToLower().Contains(keyword.ToLower())).Select(i => i.Brand.Name).Distinct().ToList();
                var totalItem = query.Count();
                return new APIRes(200, "Get all product successfully", new
                {
                    items = products,
                    totalItem,
                    maxPrice,
                    brands
                });
            }
            catch (System.Exception)
            {
                throw;
            }
        }
        public static APIRes getAllCategory()
        {
            try
            {
                using var dbContext = new ShieldtifyContext();
                var categories = dbContext.ItemCategories.Select(c => new
                {
                    uid = c.Uid,
                    name = c.Name,
                    description = c.Description
                }).ToList();
                return new APIRes(200, "Get all category successfully", categories);
            }
            catch (Exception)
            {
                throw;
            }
        }

        public static APIRes getAllProductByCategory(string category, int page = 1, string sort = "popular", string priceRange = "", string brandsFilter = "", string keyword = "")
        {
            try
            {
                using var dbContext = new ShieldtifyContext();
                var description = "";
                if (!Guid.TryParse(category, out Guid id))
                {
                    var categoryObj = dbContext.ItemCategories.FirstOrDefault(c => c.Name == category);
                    if (categoryObj == null)
                    {
                        return new APIRes(404, "Category not found");
                    }
                    category = categoryObj.Uid.ToString();
                    description = categoryObj.Description;
                }
                var query = dbContext.Items.Where(i => i.Categoryid.ToString() == category && i.Name.ToLower().Contains(keyword.ToLower()));
                if (priceRange != "")
                {
                    var priceRangeArr = priceRange.Split('-');
                    query = query.Where(i => i.Price >= int.Parse(priceRangeArr[0]) && i.Price <= int.Parse(priceRangeArr[1]));
                }
                if (brandsFilter != "")
                {
                    var brandsArr = brandsFilter.Split(',');
                    query = query.Where(i => brandsArr.Contains(i.Brand.Name));
                }
                if (sort == "popular")
                {
                    query = query.OrderByDescending(i => i.OrderItems.Count());
                }
                else
                {
                    var sortField = sort.Split('-')[0];
                    var sortOrder = sort.Split('-')[1];
                    if (sortField == "price")
                    {
                        query = sortOrder == "asc" ? query.OrderBy(i => i.Price) : query.OrderByDescending(i => i.Price);
                    }
                    else if (sortField == "name")
                    {
                        query = sortOrder == "asc" ? query.OrderBy(i => i.Name) : query.OrderByDescending(i => i.Name);
                    }
                }
                var products = query.Select(i => new
                {
                    uid = i.Uid,
                    name = i.Name,
                    price = i.Price,
                    primary_img = i.ItemImgs.Where(im => im.IsPrimary).Select(im => im.Link).FirstOrDefault(),
                    brand = i.Brand.Name,
                    stock_qty = i.StockQty
                }).Skip((page - 1) * 16).Take(16).ToList();
                // round up max price
                var maxPrice = (int)Math.Ceiling(dbContext.Items.Where(i => i.Categoryid.ToString() == category).Max(i => i.Price));
                var brands = dbContext.Items.Where(i => i.Categoryid.ToString() == category).Select(i => i.Brand.Name).Distinct().ToList();
                var totalItem = query.Count();
                return new APIRes(200, "Get all product by category successfully", new
                {
                    description,
                    items = products,
                    totalItem,
                    maxPrice,
                    brands
                });
            }
            catch (Exception)
            {
                throw;
            }
        }

        public static APIRes getProductDetail(string id)
        {
            try
            {
                using var dbContext = new ShieldtifyContext();
                // check id is uuid
                if (!Guid.TryParse(id, out Guid uid))
                {
                    return new APIRes(400, "Invalid product id");
                }
                var item = dbContext.Items.Where(i => i.Uid == uid)
                    .Select(i => new
                    {
                        uid = i.Uid,
                        name = i.Name,
                        price = i.Price,
                        description = i.Description,
                        specification = i.Specification,
                        imgs = i.ItemImgs.Select(im => new
                        {
                            link = im.Link,
                            is_primary = im.IsPrimary
                        }).ToList(),
                        brand = new { name = i.Brand.Name },
                        item_category = new { name = i.Category.Name },
                        stock_qty = i.StockQty
                    })
                    .FirstOrDefault();
                if (item == null)
                    return new APIRes(404, "Product not found");
                return new APIRes(200, "Get product detail successfully", item);
            }
            catch (Exception)
            {
                throw;
            }
        }
        public static APIRes createProduct(CreateProductBody body)
        {
            try
            {
                using var dbContext = new ShieldtifyContext();
                var item = new Item
                {
                    Uid = Guid.NewGuid(),
                    Categoryid = Guid.Parse(body.categoryid),
                    Brandid = Guid.Parse(body.brandid),
                    Name = body.name,
                    Specification = body.specification,
                    Description = body.description,
                    Price = body.price,
                    StockQty = body.stock_qty
                };
                dbContext.Items.Add(item);
                dbContext.SaveChanges();
                return new APIRes(200, "Create product successfully", item);
            }
            catch (Exception)
            {
                throw;
            }
        }
        public static APIRes updateProduct(string uid, UpdateProductBody body)
        {
            try
            {
                using var dbContext = new ShieldtifyContext();
                var item = dbContext.Items.FirstOrDefault(i => i.Uid.ToString() == uid);
                if (item == null)
                    return new APIRes(404, "Product not found");
                item.Categoryid = body.categoryid != null ? Guid.Parse(body.categoryid) : item.Categoryid;
                item.Brandid = body.brandid != null ? Guid.Parse(body.brandid) : item.Brandid;
                item.Name = body.name != null ? body.name : item.Name;
                item.Specification = body.specification != null ? body.specification : item.Specification;
                item.Description = body.description != null ? body.description : item.Description;
                item.Price = body.price ?? item.Price;
                item.StockQty = (int)(body.stock_qty != null ? body.stock_qty : item.StockQty);
                dbContext.SaveChanges();
                return new APIRes(200, "Update product successfully", item);
            }
            catch (Exception)
            {
                throw;
            }
        }
        public static APIRes getAllBrand()
        {
            try
            {
                using var dbContext = new ShieldtifyContext();
                var brands = dbContext.Brands.Select(b => new
                {
                    uid = b.Uid,
                    name = b.Name
                }).ToList();
                return new APIRes(200, "Get all brand successfully", brands);
            }
            catch (Exception)
            {
                throw;
            }
        }
        public static APIRes createBrand(CreateBrandBody body)
        {
            try
            {
                using var dbContext = new ShieldtifyContext();
                var brand = new Brand
                {
                    Uid = Guid.NewGuid(),
                    Name = body.name
                };
                dbContext.Brands.Add(brand);
                dbContext.SaveChanges();
                return new APIRes(200, "Create brand successfully", brand);
            }
            catch (Exception)
            {
                throw;
            }
        }
        public static APIRes addImagesToProduct(string itemid, AddImagesToProductBody body)
        {
            try
            {
                using var dbContext = new ShieldtifyContext();
                var item = dbContext.Items.FirstOrDefault(i => i.Uid.ToString() == itemid);
                if (item == null)
                    return new APIRes(404, "Product not found");
                foreach (var img in body.imgs)
                {
                    var itemImg = new ItemImg
                    {
                        Uid = Guid.NewGuid(),
                        Itemid = Guid.Parse(itemid),
                        Link = img.link,
                        IsPrimary = img.isPrimary
                    };
                    dbContext.ItemImgs.Add(itemImg);
                }
                dbContext.SaveChanges();
                return new APIRes(200, "Add images to product successfully");
            }
            catch (Exception)
            {
                throw;
            }
        }
        public static APIRes deleteImageFromProduct(string itemid, string uid)
        {
            try
            {
                using var dbContext = new ShieldtifyContext();
                var item = dbContext.Items.FirstOrDefault(i => i.Uid.ToString() == itemid);
                if (item == null)
                    return new APIRes(404, "Product not found");
                var img = dbContext.ItemImgs.FirstOrDefault(im => im.Uid.ToString() == uid);
                if (img == null)
                    return new APIRes(404, "Image not found");
                if (img.Itemid.ToString() != itemid)
                    return new APIRes(400, "Image not belong to product");
                if (img.IsPrimary)
                    return new APIRes(400, "Choose another primary image first to delete this image");
                dbContext.ItemImgs.Remove(img);
                dbContext.SaveChanges();
                return new APIRes(200, "Delete image from product successfully");
            }
            catch (Exception)
            {
                throw;
            }
        }
        public static APIRes setDefaultImage(string itemid, string uid)
        {
            try
            {
                using var dbContext = new ShieldtifyContext();
                var item = dbContext.Items.FirstOrDefault(i => i.Uid.ToString() == itemid);
                if (item == null)
                    return new APIRes(404, "Product not found");
                var img = dbContext.ItemImgs.FirstOrDefault(im => im.Uid.ToString() == uid);
                if (img == null)
                    return new APIRes(404, "Image not found");
                if (img.Itemid.ToString() != itemid)
                    return new APIRes(400, "Image not belong to product");
                var primaryImg = dbContext.ItemImgs.FirstOrDefault(im => im.Itemid.ToString() == itemid && im.IsPrimary);
                if (primaryImg != null)
                {
                    primaryImg.IsPrimary = false;
                    dbContext.SaveChanges();
                }
                img.IsPrimary = true;
                dbContext.SaveChanges();
                return new APIRes(200, "Set primary image successfully");
            }
            catch (Exception)
            {
                throw;
            }
        }
        public static APIRes createCategory(CreateCategoryBody body)
        {
            try
            {
                using var dbContext = new ShieldtifyContext();
                var category = new ItemCategory
                {
                    Uid = Guid.NewGuid(),
                    Name = body.name,
                    Description = body.description
                };
                dbContext.ItemCategories.Add(category);
                dbContext.SaveChanges();
                return new APIRes(200, "Create category successfully", category);
            }
            catch (Exception)
            {
                throw;
            }
        }
        public static APIRes updateCategory(string uid, UpdateCategoryBody body)
        {
            try
            {
                using var dbContext = new ShieldtifyContext();
                var category = dbContext.ItemCategories.FirstOrDefault(c => c.Uid.ToString() == uid);
                if (category == null)
                    return new APIRes(404, "Category not found");
                category.Name = body.name ?? category.Name;
                category.Description = body.description ?? category.Description;
                dbContext.SaveChanges();
                return new APIRes(200, "Update category successfully", category);
            }
            catch (Exception)
            {
                throw;
            }
        }
    }
}

File: server_nodejs/src/api/product/product.service.js
const { AppError } = require('../../common/errors/AppError');
const db = require('../../models');
const uuid = require('uuid');

module.exports = {
    getAllProduct: async (keyword, page, sort, priceRange, brandsFilter) => {
        try {
            var query = `FROM items it
                    LEFT JOIN item_imgs im ON it.uid = im.itemid AND im.is_primary = 1
                    LEFT JOIN order_items oi ON it.uid = oi.itemid
                WHERE lower(name) LIKE '%${keyword.toLowerCase()}%'`;
            if (priceRange) {
                query += ` AND price BETWEEN ${priceRange.split('-')[0]} AND ${
                    priceRange.split('-')[1]
                }`;
            }
            if (brandsFilter) {
                query += ` AND brandid IN (SELECT uid FROM brands WHERE name IN ('${brandsFilter
                    .split(',')
                    .join("','")}'))`;
            }
            const totalItem = (
                await db.sequelize.query({
                    query: `SELECT COUNT(*) cnt ${query}`,
                    type: db.Sequelize.QueryTypes.SELECT,
                })
            )[0][0].cnt;
            if (sort === 'popular')
                query +=
                    ' GROUP BY it.uid, name, price, link ORDER BY COUNT(oi.itemid) DESC';
            else {
                sortField = sort.split('-')[0];
                sortOrder = sort.split('-')[1];
                if (sortField === 'price') {
                    query += ` ORDER BY price ${sortOrder}`;
                } else if (sortField === 'name') {
                    query += ` ORDER BY name ${sortOrder}`;
                }
            }
            const items = await db.sequelize.query(
                `SELECT distinct it.uid, name, price, link primary_img
                ${query}
                LIMIT 16
                OFFSET ${(page - 1) * 16};`,
                {
                    type: db.Sequelize.QueryTypes.SELECT,
                },
            );
            const maxPrice = await db.sequelize.query(
                `SELECT MAX(price) max
                FROM items
                WHERE lower(name) LIKE '%${keyword.toLowerCase()}%'`,
                {
                    type: db.Sequelize.QueryTypes.SELECT,
                },
            );
            const brands = (
                await db.sequelize.query(
                    `SELECT DISTINCT br.name
                FROM items it
                    LEFT JOIN brands br ON it.brandid = br.uid
                WHERE lower(it.name) LIKE '%${keyword.toLowerCase()}%'`,
                    {
                        type: db.Sequelize.QueryTypes.SELECT,
                    },
                )
            ).map((brand) => brand.name);
            return {
                statusCode: 200,
                message: 'Get all product successfully',
                data: {
                    items,
                    totalItem,
                    maxPrice: Math.ceil(maxPrice[0].max),
                    brands,
                },
            };
        } catch (error) {
            throw new AppError(error.statusCode, error.message);
        }
    },
    getAllCategory: async () => {
        try {
            const categories = await db.item_category.findAll({
                attributes: ['uid', 'name', 'description'],
            });
            return {
                statusCode: 200,
                message: 'Get all category successfully',
                data: categories,
            };
        } catch (error) {
            throw new AppError(error.statusCode, error.message);
        }
    },
    getAllProductByCategory: async (
        category,
        page,
        sort,
        priceRange,
        brandsFilter,
        keyword,
    ) => {
        try {
            var description;
            if (!category.match(/^[a-f\d]{8}(-[a-f\d]{4}){4}[a-f\d]{8}$/i)) {
                const categoryObj = await db.sequelize.query(
                    `SELECT distinct uid, description FROM item_categories WHERE lower(name) like '${category.toLowerCase()}'`,
                    {
                        type: db.Sequelize.QueryTypes.SELECT,
                    },
                );
                if (!categoryObj.length)
                    throw new AppError(404, 'Category not found');
                category = categoryObj[0].uid;
                description = categoryObj[0].description;
            }
            var query = `FROM items it
                    LEFT JOIN item_imgs im ON it.uid = im.itemid AND im.is_primary = 1
                    LEFT JOIN order_items oi ON it.uid = oi.itemid
                    LEFT JOIN brands br ON it.brandid = br.uid
                WHERE categoryid = '${category}' AND lower(it.name) LIKE '%${keyword.toLowerCase()}%'`;
            if (priceRange) {
                query += ` AND price BETWEEN ${priceRange.split('-')[0]} AND ${
                    priceRange.split('-')[1]
                }`;
            }
            if (brandsFilter) {
                query += ` AND brandid IN (SELECT uid FROM brands WHERE name IN ('${brandsFilter
                    .split(',')
                    .join("','")}'))`;
            }
            const totalItem = await db.sequelize.query({
                query: `SELECT COUNT(*) cnt ${query}`,
                type: db.Sequelize.QueryTypes.SELECT,
            });
            if (sort === 'popular')
                query +=
                    ' GROUP BY it.uid, it.name, price, link ORDER BY COUNT(oi.itemid) DESC';
            else {
                sortField = sort.split('-')[0];
                sortOrder = sort.split('-')[1];
                if (sortField === 'price') {
                    query += ` ORDER BY price ${sortOrder}`;
                } else if (sortField === 'name') {
                    query += ` ORDER BY name ${sortOrder}`;
                }
            }
            const items = await db.sequelize.query(
                `SELECT distinct it.uid, it.name, price, link primary_img, stock_qty, br.name as brand
                ${query}
                LIMIT 16
                OFFSET ${(page - 1) * 16};`,
                {
                    type: db.Sequelize.QueryTypes.SELECT,
                },
            );
            const maxPrice = await db.sequelize.query(
                `SELECT MAX(price) max
                FROM items
                WHERE categoryid = '${category}' AND lower(name) LIKE '%${keyword.toLowerCase()}%'`,
                {
                    type: db.Sequelize.QueryTypes.SELECT,
                },
            );
            const brands = (
                await db.sequelize.query(
                    `SELECT DISTINCT br.name
                FROM items it
                    LEFT JOIN brands br ON it.brandid = br.uid
                WHERE categoryid = '${category}' AND lower(it.name) LIKE '%${keyword.toLowerCase()}%'`,
                    {
                        type: db.Sequelize.QueryTypes.SELECT,
                    },
                )
            ).map((brand) => brand.name);

            return {
                statusCode: 200,
                message: 'Get all product by category successfully',
                data: {
                    description,
                    items,
                    totalItem: totalItem[0][0].cnt,
                    maxPrice: Math.ceil(maxPrice[0].max),
                    brands,
                },
            };
        } catch (error) {
            throw new AppError(error.statusCode, error.message);
        }
    },
    getProductDetail: async (product) => {
        try {
            if (!product.match(/^[a-f\d]{8}(-[a-f\d]{4}){4}[a-f\d]{8}$/i)) {
                throw new AppError(400, 'Invalid product id');
            }
            const item = await db.item.findByPk(product, {
                include: [
                    {
                        model: db.item_img,
                        as: 'item_img',
                        attributes: ['uid', 'link', 'is_primary'],
                    },
                    {
                        model: db.brand,
                        attributes: ['name'],
                    },
                    {
                        model: db.item_category,
                        attributes: ['name'],
                    },
                ],
            });
            if (!item) throw new AppError(404, 'Product not found');
            item.dataValues.imgs = item.dataValues.item_img;
            delete item.dataValues.item_img;
            return {
                statusCode: 200,
                message: 'Get product detail successfully',
                data: item,
            };
        } catch (error) {
            throw new AppError(error.statusCode, error.message);
        }
    },
    createProduct: async ({
        categoryid,
        brandid,
        name,
        specification,
        description,
        price,
        stock_qty,
    }) => {
        try {
            const item = await db.item.create({
                uid: uuid.v4(),
                categoryid,
                brandid,
                name,
                specification,
                description,
                price,
                stock_qty,
            });
            return {
                statusCode: 200,
                message: 'Create product successfully',
                data: item,
            };
        } catch (error) {
            throw new AppError(error.statusCode, error.message);
        }
    },
    updateProduct: async (
        uid,
        {
            categoryid,
            brandid,
            name,
            specification,
            description,
            price,
            stock_qty,
        },
    ) => {
        try {
            const item = await db.item.findByPk(uid);
            if (!item) throw new AppError(404, 'Product not found');
            item.categoryid = categoryid != null ? categoryid : item.categoryid;
            item.brandid = brandid != null ? brandid : item.brandid;
            item.name = name != null ? name : item.name;
            item.specification =
                specification != null ? specification : item.specification;
            item.description =
                description != null ? description : item.description;
            item.price = price != null ? price : item.price;
            item.stock_qty = stock_qty != null ? stock_qty : item.stock_qty;
            await item.save();
            return {
                statusCode: 200,
                message: 'Update product successfully',
                data: item,
            };
        } catch (error) {
            throw new AppError(error.statusCode, error.message);
        }
    },
    getAllBrand: async () => {
        try {
            const brands = await db.brand.findAll({});
            return {
                statusCode: 200,
                message: 'Get all brand successfully',
                data: brands,
            };
        } catch (error) {
            throw new AppError(error.statusCode, error.message);
        }
    },
    createBrand: async ({ name }) => {
        try {
            const brand = await db.brand.create({
                uid: uuid.v4(),
                name,
            });
            return {
                statusCode: 200,
                message: 'Create brand successfully',
                data: brand,
            };
        } catch (error) {
            throw new AppError(error.statusCode, error.message);
        }
    },
    addImagesToProduct: async (itemid, { imgs }) => {
        try {
            const item = await db.item.findByPk(itemid);
            if (!item) throw new AppError(404, 'Product not found');
            for (const img of imgs) {
                await db.item_img.create({
                    uid: uuid.v4(),
                    itemid,
                    link: img.link,
                    is_primary: img.is_primary,
                });
            }
            return {
                statusCode: 200,
                message: 'Add images to product successfully',
            };
        } catch (error) {
            throw new AppError(error.statusCode, error.message);
        }
    },
    deleteImageFromProduct: async (itemid, uid) => {
        try {
            const item = await db.item.findByPk(itemid);
            if (!item) throw new AppError(404, 'Product not found');
            const img = await db.item_img.findByPk(uid);
            if (img.itemid !== itemid)
                throw new AppError(400, 'Image not belong to product');
            if (img.is_primary)
                throw new AppError(
                    400,
                    'Choose another primary image first to delete this image',
                );
            await img.destroy();
            return {
                statusCode: 200,
                message: 'Delete image from product successfully',
            };
        } catch (error) {
            throw new AppError(error.statusCode, error.message);
        }
    },
    setDefaultImage: async (itemid, uid) => {
        try {
            const item = await db.item.findByPk(itemid);
            if (!item) throw new AppError(404, 'Product not found');
            const img = await db.item_img.findByPk(uid);
            if (img.itemid !== itemid)
                throw new AppError(400, 'Image not belong to product');
            const primaryImg = await db.item_img.findOne({
                where: { itemid, is_primary: true },
            });
            if (primaryImg) {
                primaryImg.is_primary = false;
                await primaryImg.save();
            }
            img.is_primary = true;
            await img.save();
            return {
                statusCode: 200,
                message: 'Set primary image successfully',
            };
        } catch (error) {
            throw new AppError(error.statusCode, error.message);
        }
    },
    createCategory: async ({ name, description }) => {
        try {
            const category = await db.item_category.create({
                uid: uuid.v4(),
                name,
                description,
            });
            return {
                statusCode: 200,
                message: 'Create category successfully',
                data: category,
            };
        } catch (error) {
            throw new AppError(error.statusCode, error.message);
        }
    },
    updateCategory: async (uid, { name, description }) => {
        try {
            const category = await db.item_category.findByPk(uid);
            if (!category) throw new AppError(404, 'Category not found');
            category.name = name ? name : category.name;
            category.description = description
                ? description
                : category.description;
            await category.save();
            return {
                statusCode: 200,
                message: 'Update category successfully',
                data: category,
            };
        } catch (error) {
            throw new AppError(error.statusCode, error.message);
        }
    },
};
